flow/Net2.actor.cpp

```diff
@@ -1036,34 +1037,66 @@ ACTOR static Future<Void> reloadCertificatesOnChange( TLSConfig config, std::fun
 }
 #endif
 
-void Net2::initTLS() {
-	if(tlsInitialized) {
+void Net2::initTLS(ETLSInitState targetState) {
+	if(tlsInitializedState >= targetState) {
 		return;
 	}
 #ifndef TLS_DISABLED
-	auto onPolicyFailure = [this]() { this->countTLSPolicyFailures++; };
-	try {
-		sslHandshakerPool = createGenericThreadPool(FLOW_KNOBS->TLS_HANDSHAKE_THREAD_STACKSIZE);
-		for(int i = 0; i < FLOW_KNOBS->TLS_HANDSHAKE_THREADS; ++i) {
-			sslHandshakerPool->addThread(new SSLHandshakerThread());
+	// Any target state must be higher than NONE so if the current state is NONE
+	// then initialize the TLS config
+	if(tlsInitializedState == ETLSInitState::NONE) {
+		auto onPolicyFailure = [this]() { this->countTLSPolicyFailures++; };
+		try {
+			boost::asio::ssl::context newContext(boost::asio::ssl::context::tls);
+			const LoadedTLSConfig& loaded = tlsConfig.loadSync();
+			TraceEvent("Net2TLSConfig")
+				.detail("CAPath", tlsConfig.getCAPathSync())
+				.detail("CertificatePath", tlsConfig.getCertificatePathSync())
+				.detail("KeyPath", tlsConfig.getKeyPathSync())
+				.detail("HasPassword", !loaded.getPassword().empty())
+				.detail("VerifyPeers", boost::algorithm::join(loaded.getVerifyPeers(), "|"));
+			ConfigureSSLContext( tlsConfig.loadSync(), &newContext, onPolicyFailure );
+			sslContextVar.set(ReferencedObject<boost::asio::ssl::context>::from(std::move(newContext)));
+		} catch (Error& e) {
+			TraceEvent("Net2TLSInitError").error(e);
 		}
+		backgroundCertRefresh = reloadCertificatesOnChange( tlsConfig, onPolicyFailure, &sslContextVar );
+	}
+
+	// If a TLS connection is actually going to be used then start background threads if configured
+	if(targetState > ETLSInitState::CONFIG) {
+		int threadsToStart;
+		switch(targetState) {
+			case ETLSInitState::CONNECT:
+				threadsToStart = FLOW_KNOBS->TLS_CLIENT_HANDSHAKE_THREADS;
+				break;
+			case ETLSInitState::LISTEN:
+				threadsToStart = FLOW_KNOBS->TLS_SERVER_HANDSHAKE_THREADS;
+				break;
+			default:
+				threadsToStart = 0;
+		};
+		threadsToStart -= sslHandshakerThreadsStarted;
+
+		if(threadsToStart > 0) {
+			if(sslHandshakerThreadsStarted == 0) {
+				#if defined(__linux__)
+				if(mallopt(M_ARENA_MAX, FLOW_KNOBS->TLS_MALLOC_ARENA_MAX) != 1) {
```

Comment:
It's confusing to change the malloc behavior when TLS is enabled.  (Imagine a bug report like:  "RocksDB performance drops 50% with TLS enabled".)

Also, this makes it difficult to isolate the performance impact of TLS from the performance impact of the malloc knob, since you'd need to build custom binaries to run the benchmark.

Reply:
To use enough background threads for handshaking to see any tangible benefit on a reasonably sized cluster the thread count must be very high, so not changing malloc behavior is not a very practical option.  One can set the thread count knob to 0, however, which will do handshakes on the main thread, spawn no background threads, and not change the malloc arena count.

Reply:
I meant the opposite:  Always change it on the server, regardless of SSL settings, instead of never changing it, but we discussed this elsewhere, and I'm fine with the change as it is.